src/lib.rs

```diff
@@ -97,20 +98,99 @@ pub fn select_coin_bnb(
     inputs: &[OutputGroup],
     options: CoinSelectionOpt,
 ) -> Result<SelectionOutput, SelectionError> {
-    unimplemented!()
+    let mut selected_inputs : Vec<usize> =  vec![] ; 
+    let bnb_tries = 1000000 ; 
+    let bnb_selected_coin = bnb(&desce_input(inputs),&mut selected_inputs , 0 , 0 , bnb_tries , &options);
+    if bnb_selected_coin.is_none() {
+       return  select_coin_srd(inputs, options) ;
+    }else {
+        let selected_coin = bnb_selected_coin.unwrap();
+        let accumulated_value :u64= selected_coin.iter().fold(0, |acc, &i| acc + inputs[i].value);
+        let accumulated_weight : u32  = selected_coin.iter().fold(0, |acc, &i| acc + inputs[i].weight);
+        let estimated_fee = 0 ;
+        let waste = calculate_waste(inputs,&selected_inputs,&options,accumulated_value,accumulated_weight,estimated_fee,);
+        let selection_output =  SelectionOutput {
+             selected_inputs: selected_coin ,
+             waste : WasteMetric(waste) 
+        } ;
+        return Ok(selection_output);
+    }
+   
 }
 
 /// Return empty vec if no solutions are found
+// changing the selected_inputs : &[usize] -> &mut Vec<usize> 
 fn bnb(
     inputs_in_desc_value: &[(usize, OutputGroup)],
-    selected_inputs: &[usize],
-    effective_value: u64,
+    selected_inputs: &mut Vec<usize>,
+    effective_values : u64,
     depth: usize,
     bnp_tries: u32,
-    options: &CoinSelectionOpt,
-) -> Vec<usize> {
-    unimplemented!()
-}
+    options: &CoinSelectionOpt
+) -> option::Option<Vec<usize>> {
+    let target_for_match = options.target_value + (((options.base_weight as f32) * options.target_feerate).ceil() as u64) + options.cost_per_output ; // here we include the CostPerOutput  
+    let match_range = options.cost_per_input + options.cost_per_output; 
+    if effective_values > target_for_match + match_range {
+        return None;
+    } else if effective_values >= target_for_match {
+        return Some(selected_inputs.to_vec());
+    } else if bnp_tries <= 0 {
+        return None;
+    } else if depth >= inputs_in_desc_value.len() {
+        return None;
+    } else {
+        if generate_random_bool()==true 
+        {
+            // exploring the inclusion branch 
+            // first include then omit
+            let new_effective_values = effective_values + effective_value(&inputs_in_desc_value[depth].1 , options.target_feerate );
+            selected_inputs.push(inputs_in_desc_value[depth].0); 
+            let with_this = bnb(inputs_in_desc_value , selected_inputs , 
+                new_effective_values , depth+1, bnp_tries-1, options) ;
+                if with_this.is_some() 
+                {
+                    return with_this;
+                }else 
+                {
+                    selected_inputs.pop();  //poping out the selected utxo if it does not fit 
+                    let without_this = bnb(inputs_in_desc_value , selected_inputs , effective_values , depth+1, bnp_tries-1, options) ;
```

Comment:
`without_this` variable is not necessary.
It could be just `return without_this`